applications/models.py

```diff
@@ -56,7 +63,14 @@ class Meta:
                     | ~Q(approved_at=None) & ~Q(approved_by=None)
                 ),
                 name="%(app_label)s_%(class)s_both_approved_at_and_approved_by_set",
-            )
+            ),
+            models.CheckConstraint(
+                check=(
+                    Q(deleted_at=None) & Q(deleted_by=None)
+                    | ~Q(deleted_at=None) & ~Q(deleted_by=None)
+                ),
```

Comment:
suggestion: (via @inglesp) it might help readers be clear on the order of evaluation by using parentheses around the Boolean expressions.

Although looking again, maybe the separation on two lines is fine.